helpers/basic_auth/PAM/pam_auth.8
.TH pam_auth 8 "5 Sep 2003" "Squid PAM Auth"
.
.SH NAME
pam_auth - Squid PAM authentication helper
.
.SH SYNOPSIS
squid_pam_auth [-n "service name"] [-t TTL] [-o] [-1]
.
.SH DESCRIPTION
This helper allows Squid to connect to a mostly any available PAM
database to validate the user name and password of Basic HTTP
authentication.
.
.TP
.BI "-s " "service-name"
Specifies the PAM service name Squid uses, defaults to "squid"
.
.TP
.BI "-t " TTL
Enables persistent PAM connections where the connection to the PAM
database is kept open and reused for new logins. The TTL specifies
how long the connection will be kept open (in seconds).  Default is
to not keep PAM connections open. Please note that the use of
persistent PAM connections is slightly outside the PAM
specification and may not work with all PAM configurations.
.
.TP
.BI "-o"
Do not perform the PAM account management group (account
expiration etc)

.
.SH CONFIGURATION
.
The program needs a PAM service to be configured in 
.B /etc/pam.conf
or
.B /etc/pam.d/<servicename>
.P
The default service name is "squid", and the program makes use
of the
.BR "" ' auth "' and '" account '
management groups to verify the password and the accounts validity.
.P
For details on how to configure PAM services, see the PAM
documentation for your system. This manual does not cover PAM
configuration details.
.
.SH NOTES
.
When used for authenticating to local UNIX shadow password databases
the program must be running as root or else it won't have sufficient
permissions to access the user password database. Such use of this
program is not recommended, but if you absolutely need to then make
the program setuid root
.RS
.P
.B chown root pam_auth
.br
.B chmod u+s pam_auth
.RE
.P
Please note that in such configurations it is also strongly recommended
that the program is moved into a directory where normal users cannot
access it, as this mode of operation will allow any local user to
brute-force other users passwords. Also note the program has not been
fully audited and the author cannot be held responsible for any security
issues due to such installations.
.
.SH AUTHOR
Squid pam_auth and this manual is written by
.I Henrik Nordstrom <hno@squid-cache.org>
.
.SH COPYRIGHT
Squid pam_auth and this manual is Copyright 1999,2002,2003
Henrik Nordstrom <hno@squid-cache.org>
.
.SH QUESTIONS
Questions on the usage of this program can be sent to the
.I Squid Users <squid-users@squid-cache.org>
mailing list.
.
.SH REPORTING BUGS
Report bugs or bug-fixes to
.I Squid Bugs <squid-bugs@squid-cache.org>
or ideas for new improvements to 
.I Squid Developers <squid-dev@squid-cache.org>
.
.SH "SEE ALSO"
.BR pam (8), " PAM Systems Administrator Guide"


ChangeLog
Changes to squid-3.1.0.8 (24 May 2009):

	- Bug 2656: Pinger dies with general protection fault
	- Bug 2650: configure requires epoll_ctl in libepoll when --enable-epoll used
	- Bug 2648: Authentificator processes deferring and don't shutdown.
	- Bug 2645: allow squid to ignore must-revalidate
	- Bug 2644: auth scheme initialization is broken
	- Bug 2632: Make number of reforwarding tries configurable
	- Bug 2628: --with-pidfile=PATH option to override DEFAULT_PID_FILE
	- Bug 2627: HTCP Logging
	- Bug 2615: Call libecap::adapter::Service::start() when finalizing config.
	- Bug 2589: SNMP returning no data - wrong oid decoded
	- Bug 2571: Squid with IPv6 fails to start on kernel without IPv6
	- Bug 2559: Problem parsing /0 and /0.0.0.0
	- Bug 2404: WCCP in mask mode is broken
	- ... also all bugs closed by 3.0.STABLE14, 3.0.STABLE15, 3.0.STABLE16-RC1
	- Complete Interception multiple NAT support
	- Add Content-Disposition to the known headers list.
	- Make PEER_TCP_MAGIC_COUNT configurable
	- Fix pinger install location
	- Enable TPROXY v4 spoofing of CONNECT requests
	- ... and much documentation and code polishing

Changes to squid-3.1.0.7 (08 Apr 2009):

	- Fix: several issues with ident
	- Add several language translations
	- Upgrade code testing infrastructure
	- Migrate much code to build as internal libraries
	- Support gcc 4.4
	- Support doxygen 1.5.8
	- ... and much code polish to make things read easier

Changes to squid-3.1.0.6 (01 Mar 2009):

	- Regression Fix: Support HTTP/0.9 in accelerator mode
	- Bug 2601: Hack. Convert IPv4 netmasks to CIDR in IPv6-enabled mode
	- Bug 2593: Compile errors on Solaris 10
	- Bug 2591: adaptation_access does not work
	- Bug 2588: coredump in rDNS lookup
	- Bug 2526: default ALLOW when no list specified.
	- Bug 2287: Send a 505 on requests with unsupported HTTP versions
	- Bug 419: Hop by Hop headers MUST NOT be forwarded
	- Fix external_acl_type handling of SSL certificate details
	- Obsolete: dependency on nss_common.h and nss.h
	- Support libtool2
	- ... and various documentation and code polish

Changes to squid-3.1.0.5 (03 Feb 2009):

	- Bug 2583: Fixed issue in content adaptation
	- Bug 2576: Make translate target obey --disable-auto-locale
	- Bug 2571: Add DNS failover to use IPv4-only listen when IPv6 fails.
	- Bug 2563: 99+% CPU Usage on FTP URL
	- Bug 2505, 2524, 2558: fixed several issues on connection handling
	- Fix several issues in request parsing
	- Fix memory leak from logformat parsing
	- Fix various ESI build errors
	- Make configure tests use C++ instead of C
	- Drop special localhost conversion RFC violation.
	- Add Language: Arabic
	- ... and various documentation and code polish

Changes to squid-3.1.0.4 (23 Jan 2009):

	- Regression Fix: Bug 2558: rollback bug 2395 fix.
	- Bug 2555: Fixes to SNMP-MIB
	- Bug 2550: assertion comm.cc:350 !fd_table[fd].closing()
	- Bug 2547,2548: OSX compile errors (duplicate symbols and IPv6)
	- Bug 2508: comm.cc:2035 assertion fd_table[fd].closing()
	- Bug 2330: allow keep-alive+chunked; don't add max-age for no-cache
	- Polish ZPH configuration interface
	- Several Language Conversions to new auto-negotiate
	- Port from 2.7: squidclient -V and -j options for HTTP/1.1 and 0.9 testing
	- Fix: Pconn not being used when they should.
	- Fix: Fix pinger immediate shutdowns
	- Fix: Untangle CacheManager reports from log_fqdn
	- ... and all bugs fixed for 3.0.STABLE12
	- ... and many code polish and optimization fixes.

Changes to squid-3.1.0.3 (5 Dec 2008):

	- Regression Fix: StoreIOBuffer patch removed.
	- Regression Fix: build issues with 3.1.0.2 bundle
	- Security Bug 2526: default ALLOW when no list specified
	- Bug 2525: encoding error on error pages
	- Bug 2424: slow file descriptor leak
	- Bug 2527: ICAP compile error on g++ 4.3.2
	- Bug 2523: bad assertion left in from debug
	- Bug 2395: FTP Auth errors and others not displayed
	- Update squid_kerb_auth to 1.0.5
	  with better Squid integration.
	- Fix cache_peer forcedomainname= option
	- ... and many other minor fixes

Changes to squid-3.1.0.2 (9 Nov 2008):

	- Bug 2516: error page templates not properly installed
	- Bug 2500: Solaris build issues
	- Fixes FreeBSD build issues
	- Release Notes completed
	- Languages: new Russian, Japanese, Chinese, and general updates
	- ... and other minor fixes

Changes to squid-3.1.0.1 (27 Oct 2008):

	- Bundled ntlm_auth helper renamed (see Release Notes before changing anything)
	- peername ACL added for matching against a named peer destination
	- configure option --with-logdir= added to select log files location
	- squid_kerb_auth helper updated to 1.0.3 release
	- Bug #740: allow external acl's to use reply headers in format
	- Bug #2379: obsolete dns_testnames option
	- Code test infrastructure expanded to configuration testing
	- Policy changes to negative_ttl, cache deny QUERY, refresh_pattern
	  to bring their defaults up to RFC 2616 requirements.
	- Large increase in RFC 2616 standard compliance (ongoing)
	- squid.conf cleanups for minimal config
	- Connection Pinning ported from 2.6 for NTLM passthru authentication
	- eCAP internal adaptation module support
	- Localization and CSS display control of error pages
	- Added semi-automatic documentation of source code
	- Added TE chunked encoding decoder to workaround broken HTTP/1.1 servers
	- HTCP improvements ported from 2.7 adding HTCP CLR requests
	- IPv6 (Internet Protocol version 6) support
	- ICMPv6 (Internet Control Message Protocol version 6) support
	- FTP agent now supports EPSV/EPRT commands
	- DNS internal resolver now supports AAAA and CNAME records
	- SNMP peer and client tables now support IPv6
	- SNMP peer table supports named peers with multiple entries per IP
	- SslBump: Squid-in-the-middle decryption and encryption of straight
	  CONNECT and transparently redirected SSL traffic, using configurable
	  client- and server-side certificates. While decrypted, the traffic
	  can be inspected using ICAP.

helpers/basic_auth/PAM/pam_auth.c
/*
 * $Id$
 *
 * PAM authenticator module for Squid.
 * Copyright (C) 1999,2002,2003 Henrik Nordstrom <hno@squid-cache.org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.
 *
 * Install instructions:
 *
 * This program authenticates users against a PAM configured authentication
 * service "squid". This allows you to authenticate Squid users to any
 * authentication source for which you have a PAM module. Commonly available
 * PAM modules includes "UNIX", RADIUS, Kerberos and SMB, but a lot of other
 * PAM modules are available from various sources.
 *
 * Example PAM configuration for standard UNIX passwd authentication:
 * /etc/pam.conf:
 *  squid auth     required /lib/security/pam_unix.so.1
 *  squid account  required /lib/security/pam_unix.so.1
 *
 * Note that some PAM modules (for example shadow password authentication)
 * requires the program to be installed suid root to gain access to the
 * user password database
 *
 * Change Log:
 *
 *   Version 2.2, 2003-11-05
 *      One shot mode is now the default mode of operation
 *      with persistent PAM connections enabled by -t option.
 *      Support for clearing the PAM_AUTHTOK attribute on
 *      persistent PAM connections.
 *
 *   Version 2.1, 2002-08-12
 *      Squid-2.5 support (URL encoded login, password strings)
 *
 *   Version 2.0, 2002-01-07
 *      One shot mode, command line options
 *	man page
 *
 *   Version 1.3, 1999-12-10
 *   	Bugfix release 1.3 to work around Solaris 2.6
 *      brokenness (not sending arguments to conversation
 *      functions)
 *
 *   Version 1.2, internal release
 *
 *   Version 1.1, 1999-05-11
 *	Initial version
 *
 * Compile this program with: gcc -o pam_auth pam_auth.c -lpam -ldl
 */

#include "util.h"

#include <stdio.h>
#include <assert.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <unistd.h>

#include <security/pam_appl.h>

#define BUFSIZE 8192


/* The default PAM service name */
#ifndef DEFAULT_SQUID_PAM_SERVICE
#define DEFAULT_SQUID_PAM_SERVICE "squid"
#endif

/* The default TTL */
#ifndef DEFAULT_SQUID_PAM_TTL
#define DEFAULT_SQUID_PAM_TTL 0
#endif

static char *password = NULL;	/* Workaround for Solaris 2.6 brokenness */

/*
 * A simple "conversation" function returning the supplied password.
 * Has a bit to much error control, but this is my first PAM application
 * so I'd rather check everything than make any mistakes. The function
 * expects a single converstation message of type PAM_PROMPT_ECHO_OFF.
 */
static int
password_conversation(int num_msg, const struct pam_message **msg, struct pam_response **resp, void *appdata_ptr)
{
    if (num_msg != 1 || msg[0]->msg_style != PAM_PROMPT_ECHO_OFF) {
        fprintf(stderr, "ERROR: Unexpected PAM converstaion '%d/%s'\n", msg[0]->msg_style, msg[0]->msg);
        return PAM_CONV_ERR;
    }
    if (!appdata_ptr) {
        /* Workaround for Solaris 2.6 where the PAM library is broken
         * and does not pass appdata_ptr to the conversation routine
         */
        appdata_ptr = password;
    }
    if (!appdata_ptr) {
        fprintf(stderr, "ERROR: No password available to password_converstation!\n");
        return PAM_CONV_ERR;
    }
    *resp = calloc(num_msg, sizeof(struct pam_response));
    if (!*resp) {
        fprintf(stderr, "ERROR: Out of memory!\n");
        return PAM_CONV_ERR;
    }
    (*resp)[0].resp = strdup((char *) appdata_ptr);
    (*resp)[0].resp_retcode = 0;

    return ((*resp)[0].resp ? PAM_SUCCESS : PAM_CONV_ERR);
}

static struct pam_conv conv = {
    &password_conversation,
    NULL
};

static void usage(char *program)
{
    fprintf(stderr, "Usage: %s [options..]\n", program);
    fprintf(stderr, " -n service_name\n");
    fprintf(stderr, "           The PAM service name (default \"%s\")\n", DEFAULT_SQUID_PAM_SERVICE);
    fprintf(stderr, " -t ttl    PAM connection ttl in seconds (default %d)\n", DEFAULT_SQUID_PAM_TTL);
    fprintf(stderr, "           during this time the same connection will be reused\n");
    fprintf(stderr, "           to authenticate all users\n");
    fprintf(stderr, " -o        Do not perform account mgmt (account expiration etc)\n");
    fprintf(stderr, " -1        Only one user authentication per PAM connection\n");
}

int
main(int argc, char *argv[])
{
    pam_handle_t *pamh = NULL;
    int retval = PAM_SUCCESS;
    char *user;
    /* char *password; */
    char buf[BUFSIZE];
    time_t pamh_created = 0;
    int ttl = DEFAULT_SQUID_PAM_TTL;
    const char *service = DEFAULT_SQUID_PAM_SERVICE;
    int no_acct_mgmt = 0;

    /* make standard output line buffered */
    setvbuf(stdout, NULL, _IOLBF, 0);

    while (1) {
        int ch = getopt(argc, argv, "1n:t:o");
        switch (ch) {
        case -1:
            goto start;
        case 'n':
            service = optarg;
            break;
        case 't':
            ttl = atoi(optarg);
            break;
        case '1':
            ttl = 0;
            break;
        case 'o':
            no_acct_mgmt = 1;
            break;
        default:
            fprintf(stderr, "Unknown getopt value '%c'\n", ch);
            usage(argv[0]);
            exit(1);
        }
    }
start:
    if (optind < argc) {
        fprintf(stderr, "Unknown option '%s'\n", argv[optind]);
        usage(argv[0]);
        exit(1);
    }

    while (fgets(buf, BUFSIZE, stdin)) {
        user = buf;
        password = strchr(buf, '\n');
        if (!password) {
            fprintf(stderr, "authenticator: Unexpected input '%s'\n", buf);
            goto error;
        }
        *password = '\0';
        password = strchr(buf, ' ');
        if (!password) {
            fprintf(stderr, "authenticator: Unexpected input '%s'\n", buf);
            goto error;
        }
        *password++ = '\0';
        rfc1738_unescape(user);
        rfc1738_unescape(password);
        conv.appdata_ptr = (char *) password;	/* from buf above. not allocated */

        if (ttl == 0) {
            /* Create PAM connection */
            retval = pam_start(service, user, &conv, &pamh);
            if (retval != PAM_SUCCESS) {
                fprintf(stderr, "ERROR: failed to create PAM authenticator\n");
                goto error;
            }
        } else if (!pamh || (time(NULL) - pamh_created) >= ttl || pamh_created > time(NULL)) {
            /* Close previous PAM connection */
            if (pamh) {
                retval = pam_end(pamh, retval);
                if (retval != PAM_SUCCESS) {
                    fprintf(stderr, "WARNING: failed to release PAM authenticator\n");
                }
                pamh = NULL;
            }
            /* Initialize persistent PAM connection */
            retval = pam_start(service, "squid@", &conv, &pamh);
            if (retval != PAM_SUCCESS) {
                fprintf(stderr, "ERROR: failed to create PAM authenticator\n");
                goto error;
            }
            pamh_created = time(NULL);
        }
        /* Authentication */
        retval = PAM_SUCCESS;
        if (ttl != 0) {
            if (retval == PAM_SUCCESS)
                retval = pam_set_item(pamh, PAM_USER, user);
            if (retval == PAM_SUCCESS)
                retval = pam_set_item(pamh, PAM_CONV, &conv);
        }
        if (retval == PAM_SUCCESS)
            retval = pam_authenticate(pamh, 0);
        if (retval == PAM_SUCCESS && !no_acct_mgmt)
            retval = pam_acct_mgmt(pamh, 0);
        if (retval == PAM_SUCCESS) {
            fprintf(stdout, "OK\n");
        } else {
error:
            fprintf(stdout, "ERR\n");
        }
        /* cleanup */
        retval = PAM_SUCCESS;
#ifdef PAM_AUTHTOK
        if (ttl != 0) {
            if (retval == PAM_SUCCESS)
                retval = pam_set_item(pamh, PAM_AUTHTOK, NULL);
        }
#endif
        if (ttl == 0 || retval != PAM_SUCCESS) {
            retval = pam_end(pamh, retval);
            if (retval != PAM_SUCCESS) {
                fprintf(stderr, "WARNING: failed to release PAM authenticator\n");
            }
            pamh = NULL;
        }
    }

    if (pamh) {
        retval = pam_end(pamh, retval);
        if (retval != PAM_SUCCESS) {
            pamh = NULL;
            fprintf(stderr, "ERROR: failed to release PAM authenticator\n");
        }
    }
    return 0;
}
